ocaml/xapi/xapi_observer.ml

```diff
@@ -0,0 +1,211 @@
+ * GNU Lesser General Public License for more details.
+ *)
+let supported_components = ["xapi"]
+
+module RefSet = Set.Make (struct
+  type t = [`host] Ref.t
+
+  let compare = Ref.compare
+end)
+
+let observed_hosts_of ~__context hosts =
+  match hosts with [] -> Db.Host.get_all ~__context | hosts -> hosts
+
+let assert_valid_hosts ~__context hosts =
+  List.iter
+    (fun self ->
+      if not (Db.is_valid_ref __context self) then
+        raise
+          Api_errors.(
+            Server_error (invalid_value, ["host"; Ref.string_of self])
+          )
+    )
+    hosts
+
+let assert_valid_components components =
+  List.iter
+    (fun component ->
+      if not (List.mem component supported_components) then
+        raise
+          Api_errors.(Server_error (invalid_value, ["component"; component]))
+    )
+    components
+
+let assert_valid_endpoints endpoints =
+  let validate_endpoint = function
+    | "bugtool" ->
+        true
+    | url -> (
+      try
+        let _ = Uri.of_string url in
+        true
+      with _ -> false
+    )
+  in
+  List.iter
+    (fun endpoint ->
+      if not (validate_endpoint endpoint) then
+        raise Api_errors.(Server_error (invalid_value, ["endpoint"; endpoint]))
+    )
+    endpoints
+
+let assert_valid_attributes attributes =
+  List.iter
+    (fun (k, v) ->
+      if not (Tracing.validate_attribute (k, v)) then
+        let kv = Printf.sprintf "%s:%s" k v in
+        raise Api_errors.(Server_error (invalid_value, ["attributes"; kv]))
+    )
+    attributes
+
+let register ~__context ~self ~host =
+  let pool = Helpers.get_pool ~__context in
+  let host_label = Db.Host.get_name_label ~__context ~self:host in
+  let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+  let pool_uuid = Db.Pool.get_uuid ~__context ~self:pool in
+  let name_label = Db.Observer.get_name_label ~__context ~self in
+  let attributes =
+    ("xs.pool.uuid", pool_uuid)
+    :: ("xs.host.name", host_label)
+    :: ("xs.host.uuid", host_uuid)
+    :: ("xs.observer.name", name_label)
+    :: List.map
+         (fun (k, v) -> ("user." ^ k, v))
+         (Db.Observer.get_attributes ~__context ~self)
+  in
+  let uuid = Db.Observer.get_uuid ~__context ~self in
+  let endpoints = Db.Observer.get_endpoints ~__context ~self in
+  let enabled = Db.Observer.get_enabled ~__context ~self in
+  Tracing.create ~uuid ~name_label ~tags:attributes ~endpoints ~processors:[]
+    ~filters:[] ~enabled ~service_name:"xapi"
+
+let create' ~__context ~name_label ~name_description ~hosts ~attributes
+    ~endpoints ~components ~enabled ~registration =
+  assert_valid_components components ;
+  assert_valid_endpoints endpoints ;
+  assert_valid_hosts ~__context hosts ;
+  assert_valid_attributes attributes ;
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
+  Db.Observer.create ~__context ~ref ~uuid ~name_label ~name_description ~hosts
+    ~enabled ~attributes ~endpoints ~components ;
+  if registration then
```

Comment:
This is getting confusing, especially since the normal code flow has `registration:false`. I don't think we need it at all.

We should have:
* `create'` be the function that does input validation and creates the DB record _only_
* `create` be the function, used by the API, that calls `create'` followed by the `Client.register` API call.

Unit tests can't make `Client` calls, but can test `create'` in isolation.

_Alternatively_, add a register function as an argument to `create'`, which is calls at the end. Then `create` passes in the whole `Helpers.call_api_functions` block from below as this argument to `create'`. And unit test can pass in something else relevant to the test (which could be `Fun.id`).

Reply:
Going with the first, we could create an alternate `create` in the unit tests which adds a `Tracing.create` call this way `create'` is used in both places just with different calls. Have added as a fixup.